### PR TITLE
Add option for CONTAINS(*, <search_term>) and CONTAINS(*, <search_ter…

### LINKED ISSUE #13495: Enhancements to full-text seach
SQL Server provides an option for full-text CONTAINS that includes all indexes defined in a full-text index (rather than specifying each index separately): CONTAIN(*, <search_term>, [LANGUAGE <language_id>]). EF Core does not currently support this use case.

Although the ideal solution would be to provide a provider-agnostic extension to define a "contains from any indexed field" extension to Linq, that would require a significant refactor of the full-text search logic. I have implemented an extension method EF.Functions.ContainsAny(...) that will replace the string field represented by the member accessor in the first parameter with *.

The expression
```
 var result = await context.Employees
       .Where(c => EF.Functions.Contains(c.Title, "Representative"))
```
will generate (as currently implemented in 2.2)
```
SELECT [c].[EmployeeID], [c].[City], [c].[Country], [c].[FirstName], [c].[ReportsTo], [c].[Title]
FROM [Employees] AS [c]
WHERE CONTAINS(c.Title, N'Representative')
```
While the (proposed new) expression
```
 var result = await context.Employees
       .Where(c => EF.Functions.ContainsAny(c.Title, "Representative"))
```
will generate (as currently implemented in 2.2)
```
SELECT [c].[EmployeeID], [c].[City], [c].[Country], [c].[FirstName], [c].[ReportsTo], [c].[Title]
FROM [Employees] AS [c]
WHERE CONTAINS(*, N'Representative')
```
This only addresses the CONTAINS case, although the FREETEXT case would be trivial to implement as well.

### PR DESCRIPTION
**Pull request by milomconsulting**
Summary of the changes
    - Add CONTAINS(*, <search_term>) handling
    - Add CONTAINS(*, <search_term>, <language> handling

    Fixes #13495
```
- [x ] Tests for the changes have been added (for bug fixes / features)
- x[ ] Code meets the expectations our engineering guidelines. https://github.com/aspnet/Home/wiki/Engineering-guidelines.

### CODE REVIEW

**Comment by milomconsulting** [comment]
I thought about that. Unfortunately, all of the CONTAINS and FREETEXT logic is trggered by a member access in the lamdba. Trying to implement the string accepting version triggers a whole different execution path that ends with an exception do to client side evaluation. Working through that proven to be more than I had time for

**Comment by smitpatel** [comment]
@ilmax , @milomconsulting - Please refer to feedback by @divega on original issue https://github.com/aspnet/EntityFrameworkCore/issues/13495

**Comment by milomconsulting** [comment]
I am working on this now.  My first priority is getting the CONTAINS(*, …) working.  Once that is implemented, though, the other cases should be pretty straight forward.

**Comment by ilmax** [comment]
Nice, really looking forward to this one!

File: src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -617,12 +617,15 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
                 var arguments
                     = methodCallExpression.Arguments
                         .Where(
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs]: Convert LINQ to for loop for better readability.

File: src/EFCore.Relational/Query/Expressions/PropertyListParameterExpression.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        /// <summary>
+        ///     Create a new instance of PropertyListParameterExpression.
+        /// </summary>
+        /// <param name="properties">The collection of properties</param>
+        public PropertyListParameterExpression(IEnumerable<Expression> properties)
+        {
+            Check.NotNull(properties, nameof(properties));
+            Properties = properties
+                .OfType<MemberExpression>()
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Expressions/PropertyListParameterExpression.cs]: Why such filter?

File: src/EFCore.Relational/Query/Expressions/PropertyListParameterExpression.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        ///     Override this method to provide logic to walk the node's children.
+        ///     A typical implementation will call visitor.Visit on each of its
+        ///     children, and if any of them change, should return a new copy of
+        ///     itself with the modified children.
+        /// </remarks>
+        protected override Expression VisitChildren(ExpressionVisitor visitor)
+        {
+            foreach (var property in Properties)
+            {
+                visitor.Visit(property);
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Expressions/PropertyListParameterExpression.cs]: If any of the property change during visit, you need to reconstruct this expression.

File: src/EFCore.Relational/Query/Expressions/PropertyListParameterExpression.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            if (obj is null)
+            {
+                return false;
+            }
+
+            return ReferenceEquals(this, obj) ? true : obj.GetType() == GetType() && Equals((PropertyListParameterExpression)obj);
+        }
+
+        private bool Equals(PropertyListParameterExpression other)
+            => Equals(Properties, other.Properties);
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Expressions/PropertyListParameterExpression.cs]: SequenceEquals comparison.

File: src/EFCore.Relational/Query/Expressions/PropertyListParameterExpression.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        ///     Returns a hash code for this object.
+        /// </summary>
+        /// <returns>
+        ///     A hash code for this object.
+        /// </returns>
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                return Properties.GetHashCode();
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Expressions/PropertyListParameterExpression.cs]: foreach to get hash code

File: src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs [hunk 6]
```diff
@@ -617,12 +617,15 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
...
-                                 && !IsNonTranslatableSubquery(e.RemoveConvert()))
+                            e => !IsNonTranslatableSubquery(e.RemoveConvert()))
                         .Select(
-                            e => (e.RemoveConvert() as ConstantExpression)?.Value is Array || e.RemoveConvert().Type == typeof(DbFunctions)
-                                ? e
-                                : Visit(e))
+                            e => e.RemoveConvert() is QuerySourceReferenceExpression qsre
+                                ? new EntityParameterExpression(qsre.ReferencedQuerySource.ItemName, qsre.ReferencedQuerySource.ItemType)
+                                : e.RemoveConvert() is NewExpression ne
+                                    ? new PropertyListParameterExpression(ne.Arguments.AsEnumerable())
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs]: You cannot pass the MemberExpression as is. You need to visit them. You are printing based on member names. What if the property is mapped to a different column name on database side. This code falls apart on that.

File: src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 7]
```diff
@@ -1851,6 +1851,49 @@ public virtual Expression VisitCase(CaseExpression caseExpression)
...
+        /// <returns>
+        ///     An Expression.
+        /// </returns>
+        public virtual Expression VisitPropertyListParameter(PropertyListParameterExpression propertyListParameterExpression)
+        {
+            var fieldName = new string[propertyListParameterExpression.Properties.Length];
+
+            var idx = 0;
+            foreach (var property in propertyListParameterExpression.Properties)
+            {
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: If you do translation write then you can use GenerateList method.

File: src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 8]
```diff
@@ -1851,6 +1851,49 @@ public virtual Expression VisitCase(CaseExpression caseExpression)
...
+        /// </summary>
+        /// <param name="entityParameterExpression"> The EntityParameterExpression expression. </param>
+        /// <returns>
+        ///     An Expression.
+        /// </returns>
+        public virtual Expression VisitEntityParameter(EntityParameterExpression entityParameterExpression)
+        {
+            Check.NotNull(entityParameterExpression, nameof(entityParameterExpression));
+
+            _relationalCommandBuilder.Append("[");
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: We don't put hardcoded `[]`. Check VisitTable method to see how we delimit the table names.

File: src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerFullTextSearchMethodCallTranslator.cs [hunk 9]
```diff
@@ -31,6 +31,16 @@ public class SqlServerFullTextSearchMethodCallTranslator : IMethodCallTranslator
                 nameof(SqlServerDbFunctionsExtensions.FreeText),
                 new[] { typeof(DbFunctions), typeof(string), typeof(string), typeof(int) });
 
+        private static readonly MethodInfo _freeTextWildcardMethodInfo
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerFullTextSearchMethodCallTranslator.cs]: Change variable names, make this `_freeTextMethodInfo` and rename existing ones to `_freeTextStringMethodInfo` and likewise for others.

**Comment by smitpatel** [comment]
Changed base branch to master. Not suitable for 2.2 release.

**Comment by smitpatel** [comment]
I am closing this pull request as it requires changes. We would revisit this once we have new query pipeline.

**Pull request closed without merging**